src/ServiceControl.Monitoring/Timings/TimingsStore.cs

```diff
@@ -0,0 +1,141 @@
+﻿namespace ServiceControl.Monitoring.Timings
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using Metrics.Raw;
+
+    public abstract class TimingsStore
+    {
+        ConcurrentDictionary<string, ConcurrentDictionary<DateTime, MeasurementInterval>> timings = 
+            new ConcurrentDictionary<string, ConcurrentDictionary<DateTime, MeasurementInterval>>();
+
+        public void Store(string endpointName, LongValueOccurrences message, DateTime now)
+        {
+            var endpointData = timings.GetOrAdd(endpointName, _ => new ConcurrentDictionary<DateTime, MeasurementInterval>());
+
+            for (var i = 0; i < message.Ticks.Length; i++)
+            {
+                var date = new DateTime(message.BaseTicks + message.Ticks[i], DateTimeKind.Utc);
+                var intervalId = date.RoundDownToNearest(IntervalSize);
+
+                endpointData.AddOrUpdate(
+                    intervalId,
+                    _ => new MeasurementInterval(1, message.Values[i]),
+                    (_, b) => b.Update(message.Values[i]));
+            }
+
+            if (endpointData.Count > 2 * NumberOfHistoricalIntervals)
+            {
+                var historySize = TimeSpan.FromTicks(IntervalSize.Ticks * NumberOfHistoricalIntervals);
+                var oldestValidInterval = now.Subtract(historySize).RoundDownToNearest(IntervalSize);
+
+                foreach (var interval in endpointData.Keys)
+                {
+                    if (interval < oldestValidInterval)
+                    {
+                        MeasurementInterval bucket;
+                        endpointData.TryRemove(interval, out bucket);
+                    }
+                }
+            }
+        }
+
+        public EndpointTimings[] GetTimings(DateTime now)
+        {
+            var result = new List<EndpointTimings>();
+            var intervals = GenerateIntervalIds(now, NumberOfHistoricalIntervals);
+
+            foreach (var endpointName in timings.Keys)
```

Comment:
With the proposal above, scanning for values is easier as the linked list of chunks provides a natural skip-list based on time. Yes, there could be overlapping chunks (from different instances) but still, you need to scan chunk start/end to qualify it for data retrieval.